### DISCUSSION TITLE
Koa-body 6.0.1 bundling Issue - Cannot find module octetstream.js

### DISCUSSION THREAD

**Discussion by keymandll** [koajs/koa#1706] [Q&A] [answered]
Hi there,

I just recently upgraded from koa-body `2.5.0` to `6.0.1` and found that when I try to upload a file to my project that is bundled with `esbuild`, it just throws an exception shown below.

```
  Error: Cannot find module '/SERVICES/testenv/plugins/octetstream.js'
  Require stack:
  - /SERVICES/testenv/index.cjs
      at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
      at Function.Module._load (node:internal/modules/cjs/loader:778:27)
      at Module.require (node:internal/modules/cjs/loader:1005:19)
      at require (node:internal/modules/cjs/helpers:102:18)
      at /SERVICES/testenv/index.cjs:19916:20
      at Array.forEach (<anonymous>)
      at new IncomingForm (/SERVICES/testenv/index.cjs:19914:37)
      at formidable (/SERVICES/testenv/index.cjs:20713:36)
      at parseWithFormidable (/SERVICES/testenv/index.cjs:20744:45)
      at /SERVICES/testenv/index.cjs:20904:79
```

I have attached an example project below that shows the issue.

[testenv.zip](https://github.com/koajs/koa/files/9933584/testenv.zip)

Anyone has any ideas how the above could be fixed?

Thanks in advance.

**Comment by keymandll** [comment, marked as answer]
I think i figured it out. Koa-body (at the moment) uses formidable 2.x, which dynamically imports the plugins based on the plugin names specified via the `enabledPlugins` option. The relevant snippet from the bundled code:

```
        this.options.enabledPlugins.forEach((pluginName) => {
          const plgName = pluginName.toLowerCase();
          this.use(require(path.join(__dirname, "plugins", `${plgName}.js`)));
        });
```

This is unfortunate, because from esbuild's perspective the plugins are non-analyzable imports, quoting:

> Bundling with esbuild only works with statically-defined imports (i.e. when the import path is a string literal). Imports that are defined at run-time (i.e. imports that depend on run-time code evaluation) are not bundled, since bundling is a compile-time operation.

Looking at formidable@v3 this issue seems to be resolved as [the plugins are statically-defined imports](https://github.com/node-formidable/formidable/blob/493ec88520c960d0d848d55a929630fde62f2aae/src/Formidable.js#L32).

If anyone else runs into this problem, the above means that if you insist on using koa-body, you'd have to wait until they decide to use formidable@v3 or you could try to switch to Webpack to bundle your code (I don't know if Webpack would work though).

**Reply by nickchampion** [reply]
@keymandll did you fix this in the end? I have the same problem, I have resolved this by coping the fomidable JS files to my dist folder, but its a complete hack, wondering if you have a better solution?

Thanks

**Reply by ad-zsolt-imre** [reply]
It's been a while ago... I vaguely remember moving away from `koa-body`. I do not remember what was the alternative, though. There are packages out there with name such as `koa-better-body`, maybe worth checking them out. Sorry I cannot be of more help. Since then I moved away from JS completely as well, lol.
